src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/SyntaxNodeExtensions.cs

```diff
@@ -189,27 +192,103 @@ public static int FullWidth(this SyntaxNode node)
             return node.FullSpan.Length;
         }
 
-        public static SyntaxNode? FindInnermostCommonNode(
-            this IEnumerable<SyntaxNode> nodes,
-            Func<SyntaxNode, bool> predicate)
+        public static SyntaxNode? FindInnermostCommonNode(this IEnumerable<SyntaxNode> nodes, Func<SyntaxNode, bool> predicate)
+            => nodes.FindInnermostCommonNode()?.FirstAncestorOrSelf(predicate);
+
+        public static SyntaxNode? FindInnermostCommonNode(this IEnumerable<SyntaxNode> nodes)
         {
-            IEnumerable<SyntaxNode>? blocks = null;
+            // Two collections we use to make this operation as efficient as possible. One is a
+            // stack of the current shared ancestor chain shared by all nodes so far.  It starts
+            // with the full ancestor chain of the first node, and can only get smaller over time.
+            // It should be log(n) with the size of the tree as it's only storing a parent chain.
+            //
+            // The second is a set with the exact same contents as the array.  It's used for O(1)
+            // lookups if a node is in the ancestor chain or not.
+
+            using var _1 = ArrayBuilder<SyntaxNode>.GetInstance(out var commonAncestorsStack);
+            using var _2 = PooledHashSet<SyntaxNode>.GetInstance(out var commonAncestorsSet);
```

Comment:
we also only pool these two collections for the entire operation.  so no intermediary collections/iterators creates